### PR TITLE
Add support for metrics to gateways

### PR DESCRIPTION
**Pull request by thisisnotashwin**
Changes proposed in this PR:
- Add support for envoy metrics on the Gateways

How I've tested this PR:
Used the following values file:
```
server:
  replicas: 1
  extraConfig: |
    {
      "ui_config": {
        "enabled": true,
        "metrics_provider": "prometheus",
        "metrics_proxy": {
          "base_url": "http://prometheus-server"
        }
      }
    }
connectInject:
  enabled: true
controller:
  enabled: true
client:
  enabled: true
  grpc: true
ui:
  enabled: true
prometheus:
  enabled: true
grafana:
  enabled: true
global:
  imageK8S: ashwinvenkatesh/consul-k8s@sha256:dddeb63f220e9b0db794809f8fab2472434af4fca224246dab50617795af5f75
  metrics:
    enabled: true
    enableAgentMetrics: true
    enableGatewayMetrics: true
ingressGateways:
  enabled: true
meshGateway:
  enabled: true
terminatingGateways:
  enabled: true
```

Once all of the pods are running, port forward `9090` (prometheus port):

image

How I expect reviewers to test this PR:
Run the above

Checklist:
- [x] Bats tests added
- [ ] CHANGELOG entry added (*HashiCorp engineers only, community PRs should not add a changelog entry*)

### CODE REVIEW

File: values.yaml [hunk 1]
```diff
@@ -245,6 +245,10 @@ global:
     # Only applicable if global.metrics.enabled is true.
     agentMetricsRetentionTime: 1m
 
+    # Configures mesh, ingress, and terminating gateways’
+    # Envoy metrics to Prometheus. Only applicable if global.metrics.enabled is true.
```
Comment (ishustava) [inline values.yaml:249]: I think it'd be good to expand here this a bit. Maybe something like:

If true, mesh, terminating, and ingress gateways will expose their Envoy metrics on port 20200 at the `/metrics` path and all gateway pods will have Prometheus scrape annotations. Only applicable if global.metrics.enabled is true.

**Review by ishustava — COMMENTED** [review]
Looks really good! I left a couple of comments inline. 

My other question outside of that is whether you think we should add acceptance tests for this?

**Comment by thisisnotashwin** [comment]
> we should add acceptance tests for this?

@ishustava yeah, that's a great question. i was contemplating that. i was trying to think through what the assertions in that case would potentially be. do we want to assert on the annotations being present along with the proxy configuration or would we want to deploy a prometheus server alongside it and validate that proxy metrics are actually getting scraped. im uncertain how we would verify metrics scraping but i could spend some time thinking about it. WDYT?

File: test/unit/ingress-gateways-deployment.bats [hunk 2]
```diff
@@ -258,6 +258,62 @@ load _helpers
...
+  local actual=$(helm template \
+      -s templates/ingress-gateways-deployment.yaml  \
+      --set 'ingressGateways.enabled=true' \
+      --set 'connectInject.enabled=true' \
+      --set 'global.metrics.enabled=true'  \
+      --set 'global.metrics.enableGatewayMetrics=true'  \
+      . | tee /dev/stderr |
+      yq -s -r '.[0].spec.template.metadata.annotations."prometheus.io/scrape"' | tee /dev/stderr)
+  [ "${actual}" = "true" ]
+}
```
[the remark this answers is not in the document]
Comment (thisisnotashwin) [reply test/unit/ingress-gateways-deployment.bats:274]: Added some tests to this effect. WDYT?

**Comment by thisisnotashwin** [comment]
@ishustava i wonder if it makes sense to have a holistic acceptance test for all of metrics? Which checks that when all of it is enabled, that annotations are configured as expected. potentially also deploy prometheus and assert against the prom API (assuming there is one we can hit)

**Comment by ishustava** [comment]
@thisisnotashwin 

Yeah, I agree an end-to-end comprehensive test might make more sense. 

Regarding assertions, one idea is, in this example of gateways, to port-forward from each of the gateways on the metrics port and path and check that we see envoy prometheus metrics in the response. We can also do a test where prometheus is deployed, but then, like you said, I'm not sure if we can ask prometheus via API for our desired metrics (we probably could).

One thing we can do is add a test just for gateways for now and then continue adding to it as more metrics features are being added (I think you did something similar for CRD acceptance tests). 

Overall, I'm pretty open to either of these whichever you think makes more sense. Don't think it blocks the PR as long as we have some plan to test it end-to-end.

File: test/unit/ingress-gateways-deployment.bats (hunk 2 above)
[the remark this answers is not in the document]
Comment (ishustava) [reply test/unit/ingress-gateways-deployment.bats:274]: Looks phenomenal, thanks 💯

**Review by ishustava — APPROVED** [review]
Awesome work, Ashwin!

**Comment by thisisnotashwin** [comment]
>  port-forward from each of the gateways on the metrics port and path and check that we see envoy prometheus metrics in the response

I like this. Let me tinker with it. It would be a good place to think about the metrics acceptance in the big picture!

**Review by ndhanushkodi — APPROVED** [review]
Awesome job Ashwin!! Saw some gateway metrics from all 3!! Is the custom consul-k8s image being used?

**Pull request merged**